kornia/augmentation/_2d/geometric/base.py

```diff
@@ -252,6 +261,8 @@ def inverse_classes(
         transform: Optional[Tensor] = None,
         **kwargs: Any,
     ) -> Tensor:
+        [cb.on_inverse_classes_start(input, params, flags, transform) for cb in self.callbacks]
+        [cb.on_inverse_classes_end(input, params, flags, transform) for cb in self.callbacks]
```

Comment:
we cannot add these start/end in a decorator call because, in some cases, we will need a "setup"  before the hooks?

Reply:
For the logging purpose, we may use a decorator. But if we want to log more things, this way can provide a better control.

There are some setups for different functions, like to compute the parameters when it is None, etc. 

I am not quite sure what is the best practice for this here. If you got some better designs, we can do it in this PR.

Reply:
Well, my first thought is to split these functions into something like

```python
def forward(...):
    ... = self._setup_forward() # compute parameter, convert anything, etc
    self._forward_start_hooks() # execute hooks before starting the method itself
    ... = self._forward() # execute the implementation itself
    self._forward_end_hooks() # execute hooks after finishing the method itself
    return ...
```

But this may overcomplicate the API, and make it pretty unclear / splited -- so I'm not sure about it



For calling the hooks itself, instead of adding these inline loops, we may have something like:
```python
def _run_callback_hooks(augmentation: Module, hook: str, *args, **kwargs) -> None:
    for cb in augmentation.callbacks:
        if not hasattr(cb, hook):
            continue

        hook_callable = getattr(cb, hook)

        if not callable(hook_callable):
            continue

        hook_callable(*args, **kwargs)
```

```diff

@@ -208,10 +218,10 @@ class _BasicAugmentationBase(Module):
 
         params, flags = self._process_kwargs_to_params_and_flags(params, self.flags, **kwargs)
 
-        [cb.on_forward_start(input, params, flags) for cb in self.callbacks]
+        _run_callback_hooks(self, "on_forward_start", input, params, flags)
         output = self.apply_func(in_tensor, params, flags)
         output = self.transform_output_tensor(output, input_shape) if self.keepdim else output
-        [cb.on_forward_end(output, params, flags) for cb in self.callbacks]
+        _run_callback_hooks(self, "on_forward_end", output, params, flags)
         return output
```

Reply:
btw, if we consider having `_<method>_start_hooks`, the setup itself can be considered a hook too

Reply:
Ah. right. We can do it with this approach. It looks cleaner.